### PR TITLE
JDK-8322854: Incorrect rematerialization of scalar replaced objects in C2

### PR DESCRIPTION
**Pull request by JohnTortugo**
Current implementation of `PhaseMacroExpand::value_from_mem` returns `return _igvn.zerocon(ft);` when it hits a sentinel while searching for a memory operation on a given slice. One of the sentinels is the memory input of the allocate node origin of the memory slice. Therefore, `value_from_mem` may return `zeroconf(ft)` if `sfpt_mem` is the same memory edge used by the Allocate node origin of the memory slice being traversed.

The scalar replacement implementation uses `value_from_mem` during creation of metadata describing object scalar replaced (see `PhaseMacroExpand::create_scalarized_object_description`). The `create_scalarized_object_description` method is also used as part of RAM optimization implementation. The RAM optimization targets Phi nodes and therefore a memory graph loop created by a _memory phi_ node is possible to seen as part of the transformation. See image below:

This pattern doesn't show up when scalarizing objects that don't participate in allocation merges. 

To fix the issue I changed the code in `value_from_mem` to instead of using the _input_ memory edge of the Allocate as a stop condition, it will now use the projection memory edge of the Allocate.

Tested locally on windows, mac and linux x86_64 with JTREG tier1-3 and didn't observe any regression.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322854](https://bugs.openjdk.org/browse/JDK-8322854): Incorrect rematerialization of scalar replaced objects in C2 (**Bug** - P2)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17562/head:pull/17562` \
`$ git checkout pull/17562`

Update a local copy of the PR: \
`$ git checkout pull/17562` \
`$ git pull https://git.openjdk.org/jdk.git pull/17562/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17562`

View PR using the GUI difftool: \
`$ git pr show -t 17562`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17562.diff">https://git.openjdk.org/jdk/pull/17562.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17562#issuecomment-1909089270)

### CODE REVIEW

**Comment by vnkozlov** [comment]
@JohnTortugo do I understand correctly that we have a loop and the Phi node we are processing is memory input to Allocation?

If I recall correctly, the only way we get to `alloc->in(Mem)` if there is no `Initialize` node  (there are such cases).  In such case `Allocation` may not have memory out projection.

Why your case see `alloc->in(Mem)`?  

What other `Phi` node's edge points to?

I am concern if you use projection memory edge of the Allocate you may miss/skip it during search and start searching unrelated path.

**Comment by TobiHartmann** [comment]
I executed some quick testing and I see failures with
- compiler/gcbarriers/TestArrayCopyWithLargeObjectAlignment.java	
- compiler/loopstripmining/TestLoadOnBackedgeWithPrec.java
- compiler/membars/TestMemBarAcquire.java

```
  668  Phi  === 49 669 670  [[ 667 818 817 816 ]]  #rawptr:BotPTR !jvms: TestLoadOnBackedgeWithPrec::j @ bci:159 (line 55)
  668  Phi  === 49 669 670  [[ 667 818 817 816 ]]  #rawptr:BotPTR !jvms: TestLoadOnBackedgeWithPrec::j @ bci:159 (line 55)
   41  safePoint_poll_tls  === 47 0 819 0 0 42 1421 820 1423 616 0 46 45 44  |1645  [[ 43 39 ]]  !jvms: TestLoadOnBackedgeWithPrec::j @ bci:190 (line 46)
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/opt/mach5/mesos/work_dir/slaves/0db9c48f-6638-40d0-9a4b-bd9cc7533eb8-S9853/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/2cd88c15-c79b-46eb-a505-4fbf2c345f82/runs/27421909-81f2-433f-811e-a8e03bb02478/workspace/open/src/hotspot/share/opto/buildOopMap.cpp:365), pid=978899, tid=978928
#  assert(false) failed: there should be an oop in OopMap instead of a live raw oop at safepoint
```

With `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation` on AArch64.

**Comment by JohnTortugo** [comment]
@vnkozlov - Thank you for letting me know about those edge cases. I'll investigate what happens in those situations. 
I created this Gist to demonstrate the problem: https://gist.github.com/JohnTortugo/2e6f183b0bf1e465dc871246b410ef4c

@TobiHartmann - I'll try and reproduce these failures locally.

**Comment by vnkozlov** [comment]
> @vnkozlov - Thank you for letting me know about those edge cases. I'll investigate what happens in those situations.
> I created this Gist to demonstrate the problem: https://gist.github.com/JohnTortugo/2e6f183b0bf1e465dc871246b410ef4c

@JohnTortugo  - Thank you for demonstration. Now I understand the issue. Yes, your suggestion is reasonable but you need to watch out for missing allocation's memory projection - you should not use `nullptr` as sentinel. May be add assert and run testing to see if we hit it.

**Comment by vnkozlov** [comment]
@JohnTortugo are you still working on it? Please, let me know when you are ready for re-review and testing.

**Comment by JohnTortugo** [comment]
@vnkozlov - yes, I'm still working on it. Unfortunately, I had to push some changes in order to run tests on one of our internal systems..

**Comment by JohnTortugo** [comment]
Hi @vnkozlov , @TobiHartmann - can you please take a look at the changes and run your internal tests on it? TIA!

@vnkozlov - I added an assert to check that there is a (non_io) memory projection in the Allocate. I ran `hotspot-all` in a few OS's and the assert didn't trigger. I kept it on the code just in case it may trigger with a different set of JVM options, OS/arch, etc.

File: test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndMemoryLoop.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndMemoryLoop.java:69]: Missing new line? Why GitHub shows (-) ?

File: test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndMemoryLoop.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndMemoryLoop.java:28]: You may also add `@requires vm.compiler2.enabled` or `vm.compMode != "Xint"`

**Review by vnkozlov — COMMENTED** [review]
Looks good. Few comments for test.
Tobias testing is still running. So far no new failure.

**Comment by vnkozlov** [comment]
We may need to backport this. See my comments in [JDK-8324739](https://bugs.openjdk.org/browse/JDK-8324739).

**Comment by vnkozlov** [comment]
@JohnTortugo can you look on the `Test.java` from [JDK-8324739](https://bugs.openjdk.org/browse/JDK-8324739)?
It passed with your initial (v00) fix and failed with latest when you added `, /*io_use:*/false`

**Comment by JohnTortugo** [comment]
Thanks for testing Vladimir. I'll take a look into this tomorrow morning PST.

**Comment by vnkozlov** [comment]
@JohnTortugo, thank you for investigating [JDK-8324739](https://bugs.openjdk.org/browse/JDK-8324739).

Let finish this fix then. Please, address my comments about test.
And we need second review/approval for these changes.

**Comment by JohnTortugo** [comment]
@TobiHartmann - did any test fail in your latest run?

**Comment by TobiHartmann** [comment]
All tests passed.

**Review by TobiHartmann — APPROVED** [review]
The fix looks reasonable to me.

**Comment by JohnTortugo** [comment]
@vnkozlov - all looking good to you?

**Review by vnkozlov — APPROVED** [review]
All look good.

**Pull request closed without merging**
